src/libstrongswan/plugins/wincapi/wincapi_private_key.c

```diff
@@ -0,0 +1,596 @@
+		buf = (BYTE*) malloc(buf_len);
+		if (!CryptSignHash(hash_obj, this->key_spec, NULL, 0, buf, &buf_len))
+		{
+			return FALSE;
+		} else
+		{
+			/* In-place byte reversal of signature */
+			for (int i = 0; i < buf_len / 2; i++)
+			{
+				unsigned char c;
+				c = buf[i];
+				buf[i] = buf[buf_len - i - 1];
+				buf[buf_len - i - 1] = c;
+			}
+		}
+	} else
+	{
+		return FALSE;
+	}
+
+	if (hash_obj)
+	{
+		CryptDestroyHash(hash_obj);
+	}
+	*signature = chunk_create(buf, buf_len);
+	return TRUE;
+}
+
+
+METHOD(private_key_t, get_type, key_type_t,
+		private_wincapi_private_key_t *this)
+{
+	return this->key_type;
+}
+
+
+METHOD(private_key_t, decrypt, bool,
+		private_wincapi_private_key_t *this, encryption_scheme_t scheme,
+		chunk_t crypto, chunk_t *plain)
+{
+	BYTE *buf;
+	DWORD buf_len;
+
+	if (scheme != ENCRYPT_RSA_PKCS1)
+	{
+		DBG1(DBG_LIB, "encryption scheme %N not supported",
+				encryption_scheme_names, scheme);
+		return FALSE;
+	}
+
+	buf_len = crypto.len;
+	buf = (BYTE*) malloc(buf_len);
+	memcpy_s(buf, buf_len, crypto.ptr, crypto.len);
+	if (!CryptDecrypt(this->key, 0, TRUE, 0, buf, &buf_len))
+	{
+		return FALSE;
+	}
+
+	*plain = chunk_create(buf, buf_len);
+	return TRUE;
+}
+
+
+METHOD(private_key_t, get_keysize, int,
+		private_wincapi_private_key_t *this)
+{
+	return this->key_len;
+}
+
+
+METHOD(private_key_t, get_public_key, public_key_t*,
+		private_wincapi_private_key_t *this)
+{
+	return this->pubkey;
+}
+
+
+METHOD(private_key_t, get_encoding, bool,
+		private_wincapi_private_key_t *this, cred_encoding_type_t type,
+		chunk_t *encoding)
+{
+	return FALSE;
+}
+
+
+METHOD(private_key_t, get_fingerprint, bool,
+		private_wincapi_private_key_t *this, cred_encoding_type_t type, chunk_t *fp)
+{
+	return this->pubkey->get_fingerprint(this->pubkey, type, fp);
+}
+
+
+METHOD(private_key_t, get_ref, private_key_t*,
+		private_wincapi_private_key_t *this)
+{
+	ref_get(&this->ref);
+	return &this->public.key;
+}
+
+
+METHOD(private_key_t, destroy, void,
+		private_wincapi_private_key_t *this)
+{
+	if (ref_put(&this->ref))
+	{
+		if (this->pubkey)
+		{
+			this->pubkey->destroy(this->pubkey);
+		}
+		if (this->key)
+		{
+			CryptDestroyKey(this->key);
+		}
+		if (this->prov)
+		{
+			CryptReleaseContext(this->prov, 0);
+		}
+		if (this->cert_ctx)
+		{
+			CertFreeCertificateContext(this->cert_ctx);
+		}
+		if (this->system_store)
+		{
+			CertCloseStore(this->system_store, CERT_CLOSE_STORE_CHECK_FLAG);
+		}
+		if (this->key_container_name)
+		{
+			free(this->key_container_name);
+		}
+	}
+}
+
+
+/**
+ * See header.
+ */
+wincapi_private_key_t *wincapi_private_key_get(key_type_t type, va_list args)
+{
+	private_wincapi_private_key_t *this;
+	identification_t *subject = NULL;
+	chunk_t data = chunk_empty;
+
+	while (TRUE)
+	{
+		builder_part_t build_part = va_arg(args, builder_part_t);
+		switch (build_part)
+		{
+			case BUILD_BLOB_PEM:
+				data = va_arg(args, chunk_t);
+				subject = identification_create_from_string(data.ptr);
```

Comment:
Hm, passing an identity in `BUILD_BLOB_PEM` is definitely not ideal. You might want to use `BUILD_PKCS11_KEYID` (and optionally the other `BUILD_PKCS11_*` parameters), which is usually used for keys that are not directly accessible. Then e.g. pass and use the SHA-1 hash of the public key for the lookup via `CERT_FIND_SHA1_HASH`.